### PR TITLE
Temperature projectiles respect insulation

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request
Guns, Spells, and mobs that use temperature based projectiles got an update so that the projectile must deal with insulation when applying temperature changes.
The more insulation you have the less body temperature change you will take.

This does not change how much damage you take from a projectile.

The projectiles have been updated from the temperature to be set as, to the amount of change they can do.
The old security temp gun would always set the body temp to 100 kelvin, now each shot tries to reduce your body temp by -50 kelvin before insulation. The reverse is also true for temp projectiles that heat you up.

This does mean that naked you will take more body temp loss, and in a hard suit you have almost no temp change. (hardsuits do protect you from space cold so this follows)

edit: this does not touch cryo sting, that uses frost oil.

## Why It's Good For The Game
Better handling of temperature shots.
Insulation matters in combat.
Cold blooded creatures are better off with good insulation.

## Changelog
:cl:
balance: Temperature based projectiles respect insulation
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
So you've kneecapped the temperature gun without fixing the actual problem here.

Ash walkers will still die to watcher blasts. Lizardpeople still get fucked over by cryo stings and brief exposure to space unless they're a powergamer who keeps a hot drink on them at all times. Lizardpeople are now an objectively terrible mining race, even though they're stereotypically miners.

[This PR was how to fix this.](https://github.com/tgstation/tgstation/pull/48889) This PR just nerfs the temperature gun to the point of near-uselessness. What, exactly, are you supposed to use the temperature gun on if this PR is merged that you can't deal with better by just using a disabler?

**Comment by RaveRadbury** [comment]
> [This PR was how to fix this.](https://github.com/tgstation/tgstation/pull/48889)
It clearly wasn't as Oranges and Cyberboss both chimed in on how they didn't think it was a good solution. These temperature values weren't set with cold blooded mobs in mind and now that we have such a feature we shouldn't compromise it. It makes more sense to re-evaluate what we're doing with temperature values and improve the system rather than an easy fix.

**Comment by nightred** [comment]
Multiple hits with the temp gun stack, after about 6-8 hits a human will be at near 0 kelvin. (same but reversed for heat guns). This change makes it so that one hit to a naked lizard will not deal 93 points of burn damage as they warm up, but only about 25 burn damage.
You put that same lizard in a hardsuit and now it is immune to the temperature change from the gun.

The gun I would say is better now, as the effect stacks for worse effect if you can keep hitting them compared to the old gun. 

Also Ash lizards are sitting in a lovely 350 kelvin lava land, watchers will actually hurt them less now. They have a hot area to warm up in, and they will have a high body heat that will take many shots by a watcher to start hurting the ash lizard.

Finally, the other fix was rejected by the maintainers, this is an alternative. frankly this is also a better fix as it is how the guns should have worked from the start in my opinion.

**Comment by improvedname** [comment]
just drink coffee litteraly you heat as fast as you cooldown

**Comment by Shadowflame909** [comment]
Yeah this is cringe

Why make the temperature damaging items useless on the things they were supposed to counter.

No no no nuke op you can't just equip your hardsuit in EVA mode to tank all our weapons. You gotta take off your hardsuit now or you croak.

Same for cryo sting on the hardsuit wearing HoS/Captain.

This is sad.

**Comment by Shadowflame909** [comment]
Yeah this is cringe

Why make the temperature damaging items useless on the things they were supposed to counter.

No no no nuke op you can't just equip your hardsuit in EVA mode to tank all our weapons. You gotta take off your hardsuit now or you croak.

Same for cryo sting on the hardsuit wearing HoS/Captain.

This is sad.

**Comment by nightred** [comment]
> Why make the temperature damaging items useless on the things they were supposed to counter.
This does not make them useless, This is a buff to overall effect by being able to lower (or raise) body temp more than the old weapons. Before the cryo gun forced you to 100 Kelvin with every shot, now the cryo gun reduces your temp by 50 degrees kelvin each time you are hit making much more powerful if you keep hitting.

> No no no nuke op you can't just equip your hardsuit in EVA mode to tank all our weapons. You gotta take off your hardsuit now or you croak.
This does not cause issues with Nukies, at no point has the only weapon against them been temperature based. (the one temp fun in the armoury) The station has limited cryo weapons, and you are not gathering watchers to fight Nukies are you.

> Same for cryo sting on the hardsuit wearing HoS/Captain.
Also why are people dying hard suits? Cryo Sting uses frost oil and I have made no change to that.

**Comment by RaveRadbury** [comment]
> Same for cryo sting on the hardsuit wearing HoS/Captain.

Cryo sting is one of the few temperature interactions that behaves as intended. A sting pierces the hardsuit and affects the body directly. The problem is that a lot of temperature reactions copied this concept and just applied temperatures _directly_ to the body without consideration for clothing. This is just a step in the development of a better temp system. Oranges has said that he wants to see overheating in hardsuits so I think we're going to find that happening in the near future.

**Comment by ATH1909** [comment]
> > [This PR was how to fix this.](https://github.com/tgstation/tgstation/pull/48889)
> 
> It clearly wasn't as Oranges and Cyberboss both chimed in on how they didn't think it was a good solution. These temperature values weren't set with cold blooded mobs in mind **and now that we have such a feature we shouldn't compromise it.** It makes more sense to re-evaluate what we're doing with temperature values and improve the system rather than an easy fix.

Making an existing, popular roundstart race cold-blooded on a whim ("because muh realism") with pretty much no compensation given to them for that massive debuff was such a fundamentally bad idea that I genuinely think that it _should_ be compromised. Sometimes, "realistic" mechanics make for really shitty gameplay. You don't immediately get violently sucked out into space and killed whenever a breach to space is made in a pressurized area you're in, for example, even though that isn't "realistic".



> Multiple hits with the temp gun stack, after about 6-8 hits a human will be at near 0 kelvin. (same but reversed for heat guns). This change makes it so that one hit to a naked lizard will not deal 93 points of burn damage as they warm up, but only about 25 burn damage.

You didn't answer my question: Why would you EVER use the temperature gun over a disabler? A disabler takes out most humanoid targets immediately after a 4th or 5th shot connects, slows down said humanoid targets IMMEDIATELY after the second hit, and can be put in your backpack.

In other words, why would I settle for a gun that I have to get permission from the Warden to use that takes out targets 20-30 seconds after I get "6-8 hits" in on them when I could instead use a gun that starts in my inventory at roundstart (and also appears in the same Armory that the temperature gun is kept in) that IMMEDIATELY takes out a target after I get 4-5 hits in on them (and can be stored put in my backpack as well)?

Also, doesn't this PR fuck over the cryokinesis mutation as well? You can't spam that like you can a temperature gun; cryokinesis bolts are gonna become basically only useful for their ability to create an ice tile if they reach the end of their range.

**Comment by nightred** [comment]
>You didn't answer my question: Why would you EVER use the temperature gun over a disabler? A disabler takes out most humanoid targets immediately after a 4th or 5th shot connects, slows down said humanoid targets IMMEDIATELY after the second hit, and can be put in your backpack.

Flavor, because it is not the best does not mean it is pointless.

As for cryokinesis bolts if it needs a buff to make it much more like the old style that would be simple tweak. And that change does make sense. I reviewed some of the other items that got touched and lined the per hit and fire rate to make more sense.

Just tested cryokinesis bolts on a lizard and three hits from this new bold killed them and they never recovered completely after each hit so it just got worse each time.

I also hit a human 6 times fast (naked) with the temp gun and it passed out in under 10 seconds with speed reduction due to temp. That sounds usable to me.

**Comment by RaveRadbury** [comment]
> Making an existing, popular roundstart race cold-blooded on a whim ("because muh realism") 

Lizards were already cold-blooded, it was just poorly expressed in code by them having a lower cold tolerance.

>with pretty much no compensation given to them for that massive debuff was such a fundamentally bad idea that I genuinely think that it _should_ be compromised. 

The original PR sought to add better temperature features to races. This is the follow up for a specific race. If you feel like lizards deserve a buff for this go get maintainer approval and do it.

**Comment by Shadowflame909** [comment]
If a pr wants to completely overhaul and change the way weapons interact with players. I think said PR should also rebalance those same weapons. Otherwise you're leaving others to fix your gaps and holes!

**Comment by nightred** [comment]
> If a pr wants to completely overhaul and change the way weapons interact with players. I think said PR should also rebalance those same weapons. Otherwise you're leaving others to fix your gaps and holes!

I commented earlier that i have gone over every temperature projectile and made some modifications to better aligned them to the intended damage.
This does not modify anything other then Temperature projectiles, no other projectile was touched. and the gun was not touched directly.

**Comment by Iatots** [comment]
Both this and the hug PR implement their features by having a flat +/- temperature effect. I'd rather it be more of a "averaging" effect, where two bodies coming in contact try to reach the same temperature. In the case of temperature bolts, they could have an innate temperature and mass value, the temperature functioning as a hard cap, and the mass as a damage multiplier against small/humanoid/large creatures.

**Comment by ExcessiveUseOfCobblestone** [comment]
>Before the cryo gun forced you to 100 Kelvin with every shot, now the cryo gun reduces your temp by 50 degrees kelvin each time you are hit making much more powerful if you keep hitting.

?

**Comment by nightred** [comment]
> > Before the cryo gun forced you to 100 Kelvin with every shot, now the cryo gun reduces your temp by 50 degrees kelvin each time you are hit making much more powerful if you keep hitting.
> 
> ?

The cryo gun did this
`L.adjust_bodytemperature(((100-blocked)/100)*(temperature - L.bodytemperature)) // the new body temperature is adjusted by 100-blocked % of the delta between body temperature and the bullet's effect temperature`
And temperature was 100 this had the \affect that anything hit would would have their body temperature set to 100 degrees kelvin.
This code does account for the insulation the person is wearing, it is a brute force temp change.
This also means every shot would set target mob to 100 degrees kelvin never lower never higher.
The block value is not used in on most of these projectiles, and would apply the full change.

This makes each hit apply -50 degrees kelvin to the body temp with each hit. The first shot is not bad, but 2-4 later and your freezing to death.

**Comment by ATH1909** [comment]
> Flavor, because it is not the best does not mean it is pointless.

"Being flavorful" and "being a cool weapon" aren't mutually exclusive, y'know.

**Comment by ExcessiveUseOfCobblestone** [comment]
ah okay
